components/identity/issuer/issuer.go

```diff
@@ -1,103 +1,332 @@
 	idenStateWriter idenstatewriter.IdenStateWriter
-	signer          idensigner.IdenSigner
+	keyStore        *keystore.KeyStore
+	kOpComp         *babyjub.PublicKeyComp
 	storage         db.Storage
+	nonceGen        *UniqueNonceGen
+	idenStateList   *StorageList
 	cfg             Config
 }
 
-// TODO
-func New(storage db.Storage, cfg Config, signer idensigner.IdenSigner, idenStateWriter idenstatewriter.IdenStateWriter) (*Issuer, error) {
+// loadMTs loads the three identity merkle trees from the storage using the configuration.
+func loadMTs(cfg *Config, storage db.Storage) (*merkletree.MerkleTree, *merkletree.MerkleTree, *merkletree.MerkleTree, error) {
 	cltStorage := storage.WithPrefix(dbPrefixClaimsTree)
 	retStorage := storage.WithPrefix(dbPrefixRevocationTree)
 	rotStorage := storage.WithPrefix(dbPrefixRootsTree)
 
 	clt, err := merkletree.NewMerkleTree(cltStorage, cfg.MaxLevelsClaimsTree)
 	if err != nil {
-		return nil, err
+		return nil, nil, nil, err
 	}
 	ret, err := merkletree.NewMerkleTree(retStorage, cfg.MaxLevelsRevocationTree)
 	if err != nil {
-		return nil, err
+		return nil, nil, nil, err
 	}
 	rot, err := merkletree.NewMerkleTree(rotStorage, cfg.MaxLevelsRootsTree)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	return clt, ret, rot, nil
+}
+
+// New creates a new Issuer, creating a new genesis ID and initializes the storages.
+func New(cfg Config, kOpComp *babyjub.PublicKeyComp, extraGenesisClaims []merkletree.Entrier, storage db.Storage, keyStore *keystore.KeyStore, idenStateWriter idenstatewriter.IdenStateWriter) (*Issuer, error) {
+	clt, ret, rot, err := loadMTs(&cfg, storage)
+	if err != nil {
+		return nil, err
+	}
+
+	tx, err := storage.NewTx()
+
+	if err != nil {
+		return nil, err
+	}
+
+	nonceGen := NewUniqueNonceGen(NewStorageValue(dbKeyNonceIdx))
+	nonceGen.Init(tx)
+
+	kOp, err := kOpComp.Decompress()
+	if err != nil {
+		return nil, err
+	}
+	nonce, err := nonceGen.Next(tx)
+	if err != nil {
+		return nil, err
+	}
+	claimKOp := claims.NewClaimAuthorizeKSignBabyJub(kOp, nonce)
+	id, _, err := genesis.CalculateIdGenesisMT(clt, claimKOp, extraGenesisClaims)
 	if err != nil {
 		return nil, err
 	}
+
+	tx.Put(dbKeyId, id[:])
+	tx.Put(dbKeyKOp, kOpComp[:])
+
+	cfgJSON, err := json.Marshal(cfg)
+	if err != nil {
+		return nil, err
+	}
+	tx.Put(dbKeyConfig, cfgJSON)
+
+	idenStateList := NewStorageList(dbPrefixIdenStateList)
+
+	issuer := &Issuer{
+		id:              id,
+		claimsMt:        clt,
+		revMt:           ret,
+		rootsMt:         rot,
+		idenStateWriter: idenStateWriter,
+		keyStore:        keyStore,
+		kOpComp:         kOpComp,
+		storage:         storage,
+		nonceGen:        nonceGen,
+		idenStateList:   idenStateList,
+		cfg:             cfg,
+	}
+
+	idenState, idenStateTreeRoots := issuer.state()
+	idenStateList.Init(tx)
+
+	if err := idenStateList.Append(tx, idenState[:], &idenStateTreeRoots); err != nil {
+		return nil, err
+	}
+
+	if err := tx.Commit(); err != nil {
```

Comment:
It's ok if the commit is not reached, but some `tx.Put(...)` have already been executed?